libraries/rush-lib/src/logic/Autoinstaller.ts

```diff
@@ -109,10 +109,15 @@ export class Autoinstaller {
       rushJsonFolder: this._rushConfiguration.rushJsonFolder
     });
 
-    if (!lastInstallFlag.isValid() || lock.dirtyWhenAcquired) {
-      // Example: ../common/autoinstallers/my-task/node_modules
-      const nodeModulesFolder: string = path.join(autoinstallerFullPath, 'node_modules');
-
+    // Example: ../common/autoinstallers/my-task/node_modules
+    const nodeModulesFolder: string = path.join(autoinstallerFullPath, RushConstants.nodeModulesFolderName);
+    const isLastInstallFlagDirty: boolean =
+      !lastInstallFlag.isValid() ||
+      !Utilities.isFileTimestampCurrent(FileSystem.getStatistics(lastInstallFlag.path).mtime, [
+        nodeModulesFolder
```

Comment:
Tools sometimes create additional files in the `node_modules` folder after installation completes.  One example is the `node_modules/.cache` folder that is created by some packages.  Creating a new file or folder under `node_modules` may reset its `mtime` (depending on the filesystem).

Maybe it would be safer to test `FileSystem.exists()` instead of `isFileTimestampCurrent()`.

Reply:
So would a best general solution that works for most cases just be to check if the node_modules folder exists? Or in the case that it exists and is empty it at least tries to re-install the dependencies?

Reply:
If we wanted to catch the case where someone did `rm -Rf node_modules/*` instead of `rm -Rf node_modules`, maybe a safe/robust approach would be for Rush itself to write a file into that folder.  Generate something like this around line 129 (if the file doesn't already exist):

**node_modules/rush-autoinstaller.log**
```
If this file is deleted, Rush will assume that the node_modules folder has been
cleaned and will reinstall it.
```

Note that the filename cannot be a hidden file (e.g. `.rush-autoinstaller.log` because when executing `rm -Rf *`, the shell does not expand `*` to include hidden files).

@william2958 I'll leave it up to you whether to implement such a feature.  As-is, your PR would already solve the original problem that we encountered, so this is somewhat extra feature creep. 😄

Reply:
Although, re-reading the original issue, this solution may conflict with the node installation issue brought up in this comment: https://github.com/microsoft/rushstack/issues/2987#issuecomment-952020145 , as the log file would still be wiped and cause the re-install of the plugins

Reply:
I think the safest option is to record the mtime of node_modules in lastinstall.flag, and if they are different, let the package manager sort it out. The package managers have their own internal checks for if node_modules is up to date that will be more comprehensive than anything we could do.

Reply:
@dmichon-msft In my example above, suppose that the CLI tool writes to the `node_modules/.cache` folder every time it is invoked.  In this case, the package manager would have to be invoked every time the CLI tool is invoked. It's not invalid, but it could be inefficient.